modules/drivers/mongo/src/metabase/driver/mongo/java_driver_wrapper.clj

```diff
@@ -0,0 +1,341 @@
+(ns metabase.driver.mongo.java-driver-wrapper
+  "This namespace is a wrapper of `mongo-java-driver`, adjusted for Metabase purposes, borrowing a lot from
+   the Monger library."
+  (:require
+   [clojure.string :as str]
+   [flatland.ordered.map :as ordered-map]
+   [java-time.api :as t]
+   [metabase.config :as config]
+   [metabase.driver.util :as driver.u]
+   [metabase.lib.metadata.protocols :as lib.metadata.protocols]
+   [metabase.models.secret :as secret]
+   [metabase.query-processor.store :as qp.store]
+   [metabase.query-processor.timezone :as qp.timezone]
+   [metabase.util :as u]
+   [metabase.util.i18n :refer [trs tru]]
+   [metabase.util.log :as log]
+   [metabase.util.ssh :as ssh])
+  (:import
+   (com.mongodb ConnectionString MongoClientSettings MongoClientSettings$Builder)
+   (com.mongodb.client FindIterable MongoClient MongoClients MongoCollection MongoDatabase)
+   (com.mongodb.connection SslSettings$Builder)))
+
+(set! *warn-on-reflection* true)
+
+(def ^:dynamic *mongo-client*
+  "Stores an instance of MongoClient bound by [[with-mongo-client]]."
+  nil)
+
+;;;; Conversions
+
+(defprotocol ConvertFromDocument
+  (from-document [input keywordize] "Converts given DBObject instance to a piece of Clojure data"))
+
+(extend-protocol ConvertFromDocument
+  nil
+  (from-document [input keywordize] input)
+
+  Object
+  (from-document [input keywordize] input)
+
+  org.bson.types.Decimal128
+  (from-document [^org.bson.types.Decimal128 input keywordize]
+    (.bigDecimalValue input))
+
+  java.util.List
+  (from-document [^java.util.List input keywordize]
+    (vec (map #(from-document % keywordize) input)))
+  
+  java.util.Date
+  (from-document [t _]
+                 (t/instant t))
+  
+  org.bson.Document
+  (from-document [input keywordize]
+                 (reduce (if keywordize
+                           (fn [m ^String k]
+                             (assoc m (keyword k) (from-document (.get input k) true)))
+                           (fn [m ^String k]
+                             (assoc m k (from-document (.get input k) false))))
+                         (ordered-map/ordered-map)
+                         (.keySet input))))
+
+(defprotocol ConvertToDocument
+  (^org.bson.Document to-document [input] 
+    "Converts given piece of Clojure data to org.bson.Document usable by java driver."))
```

Comment:
Are `String`s and `Boolean`s really `org.bson.Document`? That would be magic or worse.